foreman/data_refinery_foreman/surveyor/array_express.py

```diff
@@ -25,19 +25,29 @@ def source_type(self):
 
     def determine_pipeline(self,
                            batch: Batch,
-                           key_values: List[BatchKeyValue] = []):
+                           key_values: Dict = {}):
         # If it's a CEL file run SCAN.UPC on it.
-        if batch.raw_format == "CEL":
+        if batch.files[0].raw_format == "CEL":
             return ProcessorPipeline.AFFY_TO_PCL
         # If only processed data is available then we don't need to do
         # anything to it
-        elif batch.raw_format == batch.processed_format:
+        elif batch.files[0].raw_format == batch.files[0].processed_format:
             return ProcessorPipeline.NO_OP
         # If it's not CEL and it's not already processed then we just
         # want to download it for Jackie's grant.
         else:
             return ProcessorPipeline.NONE
 
+    def group_batches(self) -> List[List[Batch]]:
+        """Groups batches based on the download URL of their only File"""
+        groups = []
+        download_urls = {file.download_url for file in [batch.files[0] for batch in self.batches]}
+
+        for url in download_urls:
+            groups.append([batch for batch in self.batches if batch.files[0].download_url == url])
```

Comment:
This approach to group the urls is compact and seems "pythonic", but the performance may not be the best. Suppose the number of elements in `self.batches` is **b**, and the number of elements in `download_urls` is **d**, the double loop means that the time will be O(b * d). But if we build a dict whose key is download_url and the value is an array of batches, it can be done in O(b + d) time. 

Of course if b and d are both small, then this big O discussion is just academic trivia. Do you know the typical values of b and d?

Reply:
I think b and d will both be small enough that it won't be an issue, but that is a much more elegant solution so I went ahead and changed it. (I managed to do it in a single pass too). I even added a test for it because I probably should have done so originally.